### PR TITLE
Follow more HLint suggestions

### PR DESCRIPTION
**Pull request by rhendric**
**Description of the change**

We're ignoring quite a large number of HLint's warnings/suggestions, and I thought it might be a good idea to do something about that.

Here it is, one commit per line removed from `.hlint.yaml` (except for the first one). I've grouped the commits into sort-of themes and ordered them approximately from what I expect will be least to most controversial (bearing in mind that there are a few suggestions left that even I didn't think would lead to improvement). A handy reference list is below, so that each one can be quibbled about individually. Maybe it would be worth it to break some of these out into smaller PRs? Should each theme be its own PR? I don't know. Let's start talking.

Commits touching more than three code files are **bolded**.

#### :brain: The no-brainers
[`0696af28`](https://github.com/purescript/purescript/pull/4090/commits/0696af28714588cea055440a797ea7184f214473) Remove unneeded HLint ignore directives
[`a7fa8610`](https://github.com/purescript/purescript/pull/4090/commits/a7fa8610f3a3a4d32a5d53ff14aec58ad3fe612b) HLint fix: "Parse error"
**[`0fd6f53f`](https://github.com/purescript/purescript/pull/4090/commits/0fd6f53ff472b8ec83d8d5141dc3c94b9b81285a) HLint fix: "Unused LANGUAGE pragma"**

#### :robot: Boolean hijinks
[`50b475c0`](https://github.com/purescript/purescript/pull/4090/commits/50b475c0802c1f4d2bb8f7da0db9ee7436919554) HLint fix: "Redundant =="
[`b6a2aef8`](https://github.com/purescript/purescript/pull/4090/commits/b6a2aef8fc570741e8fc4e654299ff2a6f374b8f) HLint fix: "Redundant if"

#### :paperclip: Did you mean...
[`c14cb8c1`](https://github.com/purescript/purescript/pull/4090/commits/c14cb8c1132ac782cfbd413428c9162422df6603) HLint fix: "Use notElem"
[`667d083a`](https://github.com/purescript/purescript/pull/4090/commits/667d083a38c61e5e3657b9a0eb365845a8c67bb7) HLint fix: "Use String"
[`b8784430`](https://github.com/purescript/purescript/pull/4090/commits/b8784430dc8b489a25f4b9b4afac46cea6ac44bd) HLint fix: "Use isDigit"
[`9f2b72a1`](https://github.com/purescript/purescript/pull/4090/commits/9f2b72a1d8af547553734a77cb4dfd48f5e23e5b) HLint fix: "Use unless"
[`2dd2764c`](https://github.com/purescript/purescript/pull/4090/commits/2dd2764c666bacaa2a01293ad376a60f7fe93cb8) HLint fix: "Use traverse_"
[`6eef8922`](https://github.com/purescript/purescript/pull/4090/commits/6eef89222c955d58b57fb094778d3908ea2067d6) HLint fix: "Use /="
[`518c3eda`](https://github.com/purescript/purescript/pull/4090/commits/518c3edaf9dc5480c2867be183ddd2b88d02ed31) HLint fix: "Use $>"
[`cc1bd476`](https://github.com/purescript/purescript/pull/4090/commits/cc1bd476c9b80531098be497b8538f2b6fa3880c) HLint fix: "Use =<<"
[`25b5042a`](https://github.com/purescript/purescript/pull/4090/commits/25b5042a1c37b1e3af386bc601305a8a3d048b0f) HLint fix: "Use maybe"
[`9491b718`](https://github.com/purescript/purescript/pull/4090/commits/9491b718bcd16fe28eee9888d8526c433a124642) HLint fix: "Use zipWithM_"
[`da6cfd3a`](https://github.com/purescript/purescript/pull/4090/commits/da6cfd3ad5572c595931831492c9832e24254199) HLint fix: "Use fromMaybe"
[`aa0269fd`](https://github.com/purescript/purescript/pull/4090/commits/aa0269fdd32edaa5d6ffe18e45115aaf197d24a6) HLint fix: "Use gets"
[`731a33f5`](https://github.com/purescript/purescript/pull/4090/commits/731a33f5c7f9d59bc97f6dd78d0252b10e952625) HLint fix: "Use unwords"

#### :scroll: To list or not to list?
[`fbcb5516`](https://github.com/purescript/purescript/pull/4090/commits/fbcb5516eaefffd213995069fcd573c1e680b5b1) HLint fix: "Use &&"
[`82d83183`](https://github.com/purescript/purescript/pull/4090/commits/82d8318300482ae0e460d6c137c67f08648f2d4c) HLint fix: "Use ++"
[`bb6e1622`](https://github.com/purescript/purescript/pull/4090/commits/bb6e1622f091beb489df643fef737446b119a60b) HLint fix: "Use list literal pattern"
[`7ff24174`](https://github.com/purescript/purescript/pull/4090/commits/7ff24174022e0f57288bf7f26daf698ad0c7ba17) HLint fix: "Use :"

#### :thinking: Complex simplifications
[`c19cacab`](https://github.com/purescript/purescript/pull/4090/commits/c19cacab81dd55b67d6323843fa14e46dab9fa2e) HLint fix: "Use ."
[`80fd0361`](https://github.com/purescript/purescript/pull/4090/commits/80fd03614d6729333c2952062dc9ee501f6e6ef5) HLint fix: "Use Just"
[`54fd3dc0`](https://github.com/purescript/purescript/pull/4090/commits/54fd3dc075a92a8e46b624935e1acf1b6ce8491d) HLint fix: "Hoist not"
[`d2edcb24`](https://github.com/purescript/purescript/pull/4090/commits/d2edcb24f9a3b3b70ec4c296cf5b61b0e56bb02f) HLint fix: "Redundant fmap"

#### :1234: Getting it sorted
**[`7cf1c80e`](https://github.com/purescript/purescript/pull/4090/commits/7cf1c80e1af2c068269d33751053b02a07903c5e) HLint fix: "Use sortOn"**
[`f279b60a`](https://github.com/purescript/purescript/pull/4090/commits/f279b60a8a984250fa0ffc026f58399912b8167e) HLint fix: "Avoid reverse"

#### :dollar: Dollars and sense 
**[`11eb5b21`](https://github.com/purescript/purescript/pull/4090/commits/11eb5b21d44476fef6c592e9bd170dec24910ba0) HLint fix: "Redundant bracket"**
**[`9540bb99`](https://github.com/purescript/purescript/pull/4090/commits/9540bb99d419d5a9ac89e404b25c0308fdb9f015) HLint fix: "Redundant $"**
[`864346c1`](https://github.com/purescript/purescript/pull/4090/commits/864346c171bd7ba374e1ae286ce1f75dc80ee26b) HLint fix: "Move brackets to avoid $"
**[`f208a8da`](https://github.com/purescript/purescript/pull/4090/commits/f208a8daa2742dc8df9ff5be9a9f944d85172d45) HLint fix: "Use <$>"**
<s>**[`cbaf6a08`](https://github.com/purescript/purescript/pull/4090/commits/cbaf6a088f23187542a3287a13394947206dc675) HLint fix: "Redundant do"**</s> https://github.com/purescript/purescript/pull/4090#discussion_r632554187

#### :point_up: Pointlessness
**[`f506048e`](https://github.com/purescript/purescript/pull/4090/commits/f506048e22d62f27d330fdc47cb604e344ce056a) HLint fix: "Eta reduce"**
[`a8f3c712`](https://github.com/purescript/purescript/pull/4090/commits/a8f3c71258083981fcab8170c1ed22b9693bf240) HLint fix: "Use join"
[`93f66a42`](https://github.com/purescript/purescript/pull/4090/commits/93f66a429e70a673caae300078215b6dd1b1797b) HLint fix: "Use fmap"
[`a3a477b3`](https://github.com/purescript/purescript/pull/4090/commits/a3a477b36b570fecc130b571ef73099f3f937c05) HLint fix: "Use const"
**[`e3a27227`](https://github.com/purescript/purescript/pull/4090/commits/e3a272272805f88e0d5eb5443ceb03e03a32da5c) HLint fix: "Use lambda-case"**

#### :haircut: Miscellaneous style choices
**[`39d75e4d`](https://github.com/purescript/purescript/pull/4090/commits/39d75e4ddcc0ea674abdc0502ddc4dc28b0f49e7) HLint fix: "Use infix"**
**[`102a4d51`](https://github.com/purescript/purescript/pull/4090/commits/102a4d515a1ab9f7c0c7805f0df24a6843a778c1) HLint fix: "Use tuple-section"**
[`74456748`](https://github.com/purescript/purescript/pull/4090/commits/744567483ace3f76888cbc8c183d109bc8607b87) HLint fix: "Use camelCase"

Finally, some words about the also-rans:

* **"Use newtype instead of data"**
  This one wants single-constructor `data` types to be declared as `newtype`, which maybe sounds harmless enough, but in some of the cases we have, such as
  ```haskell
  data ClientOptions = ClientOptions
    { clientPort :: Network.PortNumber
    }
  ```
  I think it would just obscure intent.
* **"Fuse foldr/map"**
* **"Functor law"**
* **"Fuse mapM/map"**
  All of these are triggered by cases where, if one of the HOFs in question takes a lambda and the other a point-free expression, composition could lead to code that is less clear.
* **"Avoid lambda"**
* **"Avoid lambda using `infix`"**
* **"Use section"**
  All of these really want things like `\z -> TUnknown z b` to be replaced with ``(`TUnknown` b)``, which I don't think is an improvement in cases where the function wasn't named and typed with the intent of being used infix.
* **"Use record patterns"**
  This is the weirdest suggestion. It gets triggered by any all-wildcard constructor pattern with three or more wildcards. I think whether to use `Ctor{}` or `Ctor _ _ _` has much more to do with semantics than the number of wildcards—specifically, whether adding a field to the constructor would be a good reason to revisit this code.
* **"Reduce duplication"**
  This hint might have some value but silencing it involves substantial non-mechanical refactoring, which I just didn't want to undertake.

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [ ] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

### CODE REVIEW

**Review by f-f — APPROVED** [review]

File: app/Command/Docs/Html.hs
Comment (hdgarrood) [inline app/Command/Docs/Html.hs:136]: I don't think I agree with hlint that redundant dos are a problem. In this case, I think I'd actually prefer to leave them in, as it means that you can go back and forth between elements with exactly one child vs more than one child without having to change the line which defines the parent. Can we keep ignoring the redundant do hint please?

**Review by hdgarrood — APPROVED** [review]

File: app/Command/Docs/Html.hs
Comment (rhendric) [reply app/Command/Docs/Html.hs]: Dropped.

**Comment by JordanMartinez** [comment]
Any other updates on this PR? Seems like it could be ready to merge.

**Comment by rhendric** [comment]
I wasn't planning on doing anything more in this PR other than striking commits that maintainers object to. I do think it should probably stay open for a little while longer just in case there are more objections; this covers a lot of small changes and almost all of it is subjective, and after it's squashed and merged it'll be harder to revert individual items.

**Comment by JordanMartinez** [comment]
I fixed the merge conflict due to merging #4085. I'm not sure if that will cause CI to fail now if my code doesn't pass the new hlint requirements.

**Comment by JordanMartinez** [comment]
@rhendric Since I broke your build due to merging #4085, do you want me to submit a PR to your fork to fix it? There's only three issues that need to be fixed (below):
```
Downloading and running hlint...
lib/purescript-cst/src/Language/PureScript/CST/Convert.hs:578:25: Suggestion: Move brackets to avoid $
Found:
  (N.runOpName $ qualName op) <> argName t2
Perhaps:
  N.runOpName (qualName op) <> argName t2

src/Language/PureScript/Sugar/TypeClasses/Instances.hs:38:31: Suggestion: Move brackets to avoid $
Found:
  genText <> (pack $ show uniqueIdent)
Perhaps:
  genText <> pack (show uniqueIdent)

src/Language/PureScript/TypeChecker.hs:403:6: Warning: Redundant bracket
Found:
  (d@(TypeInstanceDeclaration sa@(ss, _)
                              ch
                              idx
                              (Right dictName)
                              deps
                              className
                              tys
                              body))
Perhaps:
  d@(TypeInstanceDeclaration sa@(ss, _)
                             ch
                             idx
                             (Right dictName)
                             deps
                             className
                             tys
                             body)

3 hints
Error: Process completed with exit code 1.
```

File: app/Main.hs
Comment (JordanMartinez) [inline app/Main.hs:28]: I think this was the only thing in terms of readability where I found it easier to understand the original code rather than the updated code.

**Comment by hdgarrood** [comment]
As @JordanMartinez points out in #4094, I think it would probably be good to get this in sooner rather than later, and I think it's probably been open long enough for people to have the opportunity to object now.

> and after it's squashed and merged it'll be harder to revert individual items.

I think we should enable the "create a merge commit" merge option then, and use that option when merging this PR. Squashing and merging is a good default, but in this case the utility of keeping these commits separate is obvious.

**Comment by JordanMartinez** [comment]
> "create a merge commit" merge option then

Sounds good to me!

How should we get CI to pass again?

**Comment by hdgarrood** [comment]
I think it would be ideal to keep the history nice and clean, so the additional fixes which became made necessary after the merge should ideally be amended into the original commit which corresponds to the relevant hlint warning.

**Comment by thomashoneyman** [comment]
I've enabled normal merge commits.

File: app/Main.hs
Comment (rhendric) [reply app/Main.hs:28]: What's the strength of this opinion—are you objecting or hoping to start a discussion without necessarily objecting?

I personally prefer the updated version—I think `join` immediately communicates that what follows is an `IO` of an `IO`—but I don't mind dropping this commit if this is an actual objection.

**Comment by rhendric** [comment]
Sorry to have taken a while to check in here; I just got back from a little vacation.

I've rebased and fixed all the conflicts in their appropriate commits. I'm ready to merge (and not squash) once the fate of `"Use join"` is decided.

File: app/Main.hs
Comment (JordanMartinez) [reply app/Main.hs:28]: Mm.... just ignore the comment for now and please merge this in as soon as you can. 

The update is only not as readable to me because I am not as familiar with the code base. If I knew what `handleParseResult` and the other functions do, then the extra 'cmd' isn't necessary. That takes a few seconds to lookup.

**Review by JordanMartinez — APPROVED** [review]

**Pull request merged**
